src/lib/build/mod.rs

```diff
@@ -1,5 +1,69 @@
 //! Actually building Idris packages.
+//!
+//! The build process contains two major phases: dependency preparation and target generating.
+//!
+//! - Dependency preparation:
+//!     In this stage, elba retrieves all sources in the resolve tree, and then build each of them
+//!     in global cache directory. The works in this phase can be excuted in parallel. When all
+//!     dependencies are ready in cache, elba will copy the direct dependency into /target/deps.
+//!     Dependency preparation does not necessarily be executed along with target generating and it
+//!     could also be used by editors (like rls).
+//!
+//! - Target generating:
+//!     In this stage, Elba builds lib target, binary, docs, benchmarks and tests, only for local package.
+//!
 
-pub mod compile;
+pub mod invoke;
 pub mod context;
-pub mod job;
+pub mod prepare;
+
+use std::fs;
+use std::path::PathBuf;
+use util::{errors::Res, lock::DirLock};
+
+/// The general "mode" of what to do
+#[derive(Clone, Copy, PartialEq, Debug, Eq, Hash)]
+pub enum CompileMode {
+    /// Typecheck a target without codegen
+    Lib,
+    /// Compile and codegen executable(s)
+    ///
+    /// This subsumes the "Bench" and "Test" modes since those are just compiling and running
+    /// executables anyway
+    Bin,
+    /// Create documentation
+    Doc,
+}
+
+#[derive(Debug)]
+pub struct Layout {
```

Comment:
One defines the layout of global cache directory and the other defines the layout of target directory. We can have only one layout if we just mirror the structure of target folder to cache folder, as you outlined before, but I am not doing so since we don't(?) want docs/ tests/ bench/ to be in cache folder.

Reply:
One solution is to have a generic `artifacts/` directory which could store both all the things you mentioned plus built libraries from different codegen backends.